### PR TITLE
Device interface for Compass and memory reduction

### PR DESCRIPTION
**Pull request by lucasdemarchi**
This ports the Compass drivers to use the device interface.

Additionally the last commits do a minor cleanup:
  - remove fields that we don't need to store, which also save a little of flash space on PX4.
  - avoid rebuilding over and over the entire tree when a single driver changes

### CODE REVIEW

**Comment by lucasdemarchi** [comment]
Failed on travis, but it seems to be a travis problem:

```
The command "Tools/scripts/build_ci.sh" exited with 0.
```

Also passing on semaphore.

File: libraries/AP_Compass/AP_Compass_AK8963.cpp [hunk 1]
```diff
@@ -114,27 +116,20 @@ AP_Compass_Backend *AP_Compass_AK8963::_detect(Compass &compass,
...
 
-    if (!_bus_sem->take(100)) {
-        hal.console->printf("AK8963: Unable to get bus semaphore");
+    if (!bus_sem || !_bus->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        hal.console->printf("AK8963: Unable to get bus semaphore\n");
         goto fail_sem;
     }
 
+    if (!_bus->configure()) {
+        hal.console->printf("AK8963: Could not configure the bus\n");
```
Comment (OXINARF) [inline libraries/AP_Compass/AP_Compass_AK8963.cpp]: Shouldn't it goto fail in this case?

File: libraries/AP_HAL/AP_HAL_Boards.h [hunk 2]
```diff
@@ -304,6 +316,10 @@
 #define HAL_BARO_DEFAULT HAL_BARO_MS5611_SPI
 #define HAL_BARO_MS5611_NAME "ms5611"
 #define HAL_COMPASS_DEFAULT HAL_COMPASS_RASPILOT
+/* HMC5843 can be an external compass */
+#define HAL_COMPASS_HMC5843_I2C_BUS 1
+#define HAL_COMPASS_HMC5843_I2C_ADDR 0x1E
+#define HAL_BARO_MS5611_I2C_ADDR 0x77
```
Comment (OXINARF) [inline libraries/AP_HAL/AP_HAL_Boards.h]: Was this baro address meant to be set?

File: libraries/AP_Compass/AP_Compass_AK8963.cpp (hunk 1 above)
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_AK8963.cpp]: yep, thanks.

File: libraries/AP_HAL/AP_HAL_Boards.h (hunk 2 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL/AP_HAL_Boards.h]: no, thanks.

**Comment by lucasdemarchi** [comment]
@OXINARF Thanks! I fixed those and force pushed.

**Comment by OXINARF** [comment]
@lucasdemarchi Such a big log that I was missing where it was failing but just found it:

$ python Tools/autotest/param_metadata/param_parse.py
Unable to handle multiple vehicles with .pde library
Unable to handle multiple vehicles with .pde library
Path ../libraries/AP_Compass/Compass.cpp not found for library COMPASS_
Unable to handle multiple vehicles with .pde library
Unable to handle multiple vehicles with .pde library
Unable to handle multiple vehicles with .pde library
Unable to handle multiple vehicles with .pde library

The command "python Tools/autotest/param_metadata/param_parse.py" exited with 1."

**Comment by yanndupuy** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  dupuyyann@gmail.com
    Domain zyapps.com has exceeded the max emails per hour (400/400 (100%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext1.iad.github.net ([192.30.252.192]:49250 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by serveur.ex2-p10.com with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.86_1)
	(envelope-from <noreply@github.com>)
	id 1afmLe-000aX9-66
	for yann.dupuy@zyapps.com; Tue, 15 Mar 2016 11:37:17 +0100
Date: Mon, 14 Mar 2016 18:10:53 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1458004253;
	bh=BSF5WyQtFO7Jz05kvI2CmnAU7Knx+XrcHk5JZYV5sJg=;
	h=From:Reply-To:To:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=tUyxRySoDLe64YUWPojFrmTCPX+NRxeNYE1hjyLU+jbHdNSYUkugch5oYHWitlr4D
	 M9GVUMMyUb/nYm0HRzt2cWR5KI3pQhtTVGojsq6mbYZwKnC5L1uIB60tfO6fKef9Fd
	 nQNkJUTxsWl+7uOZFpChPAPZWF2rVaUtim0kQSnQ=

**Comment by lucasdemarchi** [comment]
@OXINARF thanks, I'll fix that.

File: libraries/AP_HAL/AP_HAL_Boards.h [hunk 3]
```diff
@@ -382,6 +406,9 @@
 #define HAL_BARO_DEFAULT          HAL_BARO_MS5611_SPI
 #define HAL_BARO_MS5611_NAME "ms5611"
 #define HAL_COMPASS_DEFAULT       HAL_COMPASS_AK8963_MPU9250
+/* HMC5843 can be an external compass */
+#define HAL_COMPASS_HMC5843_I2C_BUS 1
+#define HAL_COMPASS_HMC5843_I2C_ADDR 0x1E
```
Comment (guludo) [inline libraries/AP_HAL/AP_HAL_Boards.h]: These two values are very common. Couldn't we use them as default?
```c++
#ifndef HAL_COMPASS_HMC5843_I2C_BUS
#define HAL_COMPASS_HMC5843_I2C_BUS 1
#endif
#ifndef HAL_COMPASS_HMC5843_I2C_ADDR
#define HAL_COMPASS_HMC5843_I2C_ADDR 0x1E
#endif

File: libraries/AP_Compass/AP_Compass_HMC5843.cpp [hunk 4]
```diff
@@ -184,110 +196,177 @@ void AP_Compass_HMC5843::accumulate(void)
...
+    }
+
+    if (!_bus->block_read(REG_DATA_OUTPUT_X_MSB, (uint8_t *) &val, sizeof(val))){
+        _retry_time = AP_HAL::millis() + 1000;
+        return false;
+    }
+
+    rx = be16toh(val.rx);
+    ry = be16toh(val.ry);
+    rz = be16toh(val.rz);
```
Comment (guludo) [inline libraries/AP_Compass/AP_Compass_HMC5843.cpp]: Nice.

File: libraries/AP_Compass/Compass.cpp [hunk 5]
```diff
@@ -424,45 +424,41 @@ void Compass::_detect_backends(void)
...
+#elif CONFIG_HAL_BOARD == HAL_BOARD_LINUX && CONFIG_HAL_BOARD_SUBTYPE != HAL_BOARD_SUBTYPE_LINUX_NONE
+    _add_backend(AP_Compass_HMC5843::probe(*this, hal.i2c_mgr->get_device(HAL_COMPASS_HMC5843_I2C_BUS, HAL_COMPASS_HMC5843_I2C_ADDR)));
     _add_backend(AP_Compass_AK8963::detect_mpu9250(*this, 0));
-#elif HAL_COMPASS_DEFAULT == HAL_COMPASS_QFLIGHT
-    _add_backend(AP_Compass_QFLIGHT::detect(*this));
-#elif HAL_COMPASS_DEFAULT == HAL_COMPASS_QURT
-    _add_backend(AP_Compass_QURT::detect(*this));
 #else
     #error Unrecognised HAL_COMPASS_TYPE setting
 #endif
```
Comment (guludo) [inline libraries/AP_Compass/Compass.cpp]: Some comments:
 - I think it would be easier to review if this commit was split.
 - It would be nice to have an explanation on why not using _retry_time on general register writes and reads.
 - Have you considered to have a device class for an AuxiliaryBus rather than keeping the internal bus classes?

It looks good though.

File: libraries/AP_Compass/AP_Compass_HMC5843.cpp [hunk 6]
```diff
@@ -282,12 +303,11 @@ void AP_Compass_HMC5843::read()
     publish_filtered_field(field, _compass_instance);
 }
 
-
-bool AP_Compass_HMC5843::_exit_test_mode()
+bool AP_Compass_HMC5843::_setup_mode()
```
Comment (guludo) [inline libraries/AP_Compass/AP_Compass_HMC5843.cpp]: Why this name? Isn't it the opposite?

File: libraries/AP_Compass/AP_Compass_HMC5843.cpp [hunk 7]
```diff
@@ -360,39 +385,46 @@ bool AP_Compass_HMC5843::_detect_version()
...
+    uint8_t old_config = _base_config & ~(HMC5843_OPMODE_MASK);
+
     while (success == 0 && numAttempts < 25 && good_count < 5) {
         numAttempts++;
 
         // force positiveBias (compass should return 715 for all channels)
-        if (!_bus->register_write(ConfigRegA, PositiveBiasConfig))
-            continue;   // compass not responding on the bus
+        if (!_bus->register_write(HMC5843_REG_CONFIG_A,
+                                  old_config | HMC5843_OPMODE_POSITIVE_BIAS)) {
```
Comment (guludo) [inline libraries/AP_Compass/AP_Compass_HMC5843.cpp]: The use of `old_config` is rather a fix. Maybe a different commit for this?

File: libraries/AP_Compass/AP_Compass.cpp [hunk 8]
```diff
@@ -1,8 +1,16 @@
...
+#include "AP_Compass_HIL.h"
+#include "AP_Compass_HMC5843.h"
+#include "AP_Compass_PX4.h"
+#include "AP_Compass_QURT.h"
+#include "AP_Compass_qflight.h"
+#include "AP_Compass.h"
+
 extern AP_HAL::HAL& hal;
 
 #if APM_BUILD_TYPE(APM_BUILD_ArduCopter)
```
Comment (guludo) [inline libraries/AP_Compass/AP_Compass.cpp]: I'd prefer the files renaming to be in a separate commit.

Comment (guludo) [inline libraries/AP_Compass/AP_Compass.cpp]: Also, if we are renaming the files, shouldn't we rename the class as well?

**Comment by guludo** [comment]
@lucasdemarchi I just finished reviewing this. For some reason, GitHub thinks I've commented on outdated diffs, which is wrong hehe

File: libraries/AP_HAL/AP_HAL_Boards.h (hunk 3 above)
Comment (lucasdemarchi) [inline libraries/AP_HAL/AP_HAL_Boards.h]: I just don't want to invest time on this header... still waiting a good soul to add a reasonable config implementation that will get away with it. Meanwhile things get added in the easiest possible way.

I'll eventually get to it and add it, but first I need to finish the pending work I'm doing.

File: libraries/AP_Compass/Compass.cpp (hunk 5 above)
Comment (lucasdemarchi) [inline libraries/AP_Compass/Compass.cpp]: > I think it would be easier to review if this commit was split.

github pretends to fix its bad UI for review by making a worse one. Now I don't know what commit you are talking about.

> It would be nice to have an explanation on why not using _retry_time on general register writes and reads.

Retry in general is a bad design.  It should be way up in the decision making, not in the lower level code.  In this particular case I moved the retry to be done only while collecting samples.  If you can't pass the initialization phase without resorting to "retry", go home and fix your cabling/sensors - and... reducing the bus speed affects **all** users in that bus and can be an even worse problem than just setting the sensor as unhealthy.

Comment (lucasdemarchi) [inline libraries/AP_Compass/Compass.cpp]: > Have you considered to have a device class for an AuxiliaryBus rather than keeping the internal bus classes?

AuxiliaryBus access is very different from the Device interface. Every time I touch its code I'm tempted on writing an interface that could be used for both though. However I think it's out of scope for this PR.

Comment (lucasdemarchi) [inline libraries/AP_Compass/Compass.cpp]: > I think it would be easier to review if this commit was split.

I almost rewrote the code completely - to convert to the new API, make it cleaner and some changes just to make it more similar with other drivers. I'll take a look on splitting it, but the changes are so intricate that I consider this case to be the one in which it's easier to review the final result than what changed.

File: libraries/AP_Compass/AP_Compass_HMC5843.cpp (hunk 6 above)
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass_HMC5843.cpp]: nops... it sets up the **mode** register so the sensor is put in the normal sampling mode rather than the calibration mode.

File: libraries/AP_Compass/AP_Compass.cpp (hunk 8 above)
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass.cpp]: class would need to be renamed throughout the codebase. I prefer to do it in steps and postpone the rename when this code gets in.

File: libraries/AP_Compass/Compass.cpp (hunk 5 above)
Comment (OXINARF) [inline libraries/AP_Compass/Compass.cpp]: > github pretends to fix its bad UI for review by making a worse one. Now I don't know what commit you are talking about.

If you click the _View full outdated diff_ text at the right it takes you to the right commit

Comment (lucasdemarchi) [inline libraries/AP_Compass/Compass.cpp]: @OXINARF Thanks!

That's indeed an improvement.

Now... if only it didn't change the commit order in a PR I would be super happy :)

Comment (guludo) [inline libraries/AP_Compass/Compass.cpp]: > AuxiliaryBus access is very different from the Device interface.

As far as I remember, AuxiliaryBus has only been used to access slave *devices* so far. Is there other use for it rather than accessing devices? If so, we can think of a different approach in the future, like having a class for INS slave devices for example.

Comment (guludo) [inline libraries/AP_Compass/Compass.cpp]: > Retry in general is a bad design. It should be way up in the decision making, not in the lower level code. In this particular case I moved the retry to be done only while collecting samples. If you can't pass the initialization phase without resorting to "retry", go home and fix your cabling/sensors - and... reducing the bus speed affects all users in that bus and can be an even worse problem than just setting the sensor as unhealthy.

Thanks for the explanation :-) I think a little note on the commit message would be nice. *Your call* though.

File: libraries/AP_HAL/AP_HAL_Boards.h (hunk 3 above)
Comment (guludo) [inline libraries/AP_HAL/AP_HAL_Boards.h]: > still waiting a good soul to add a reasonable config implementation that will get away with it.

I don't think I'm a good soul, but I'd like to implement it if I have a chance :-)

File: libraries/AP_Compass/AP_Compass.cpp (hunk 8 above)
Comment (guludo) [inline libraries/AP_Compass/AP_Compass.cpp]: Now that I come to think of it, yes. I think it makes sense to postpone the rename, which would be an easier patch (PR) to review and consequently quicker to get in. If that was done here, it could conflict with new patches in master while this PR is pending.

File: libraries/AP_Compass/AP_Compass_HMC5843.cpp (hunk 6 above)
Comment (guludo) [inline libraries/AP_Compass/AP_Compass_HMC5843.cpp]: I get it now. Well, I think the function name is ambiguous. For me it gave me the idea it was going to enter in some sort of *setup mode*, whereas what the code does is quite different. Maybe rename it to `_sampling_mode` or `_setup_sampling_mode`?

File: libraries/AP_Compass/AP_Compass.cpp (hunk 8 above)
Comment (lucasdemarchi) [inline libraries/AP_Compass/AP_Compass.cpp]: > I'd prefer the files renaming to be in a separate commit.

done

**Comment by lucasdemarchi** [comment]
@guludo could you take another look?  Just updated the PR. Changes:

  - commits split and methods renamed as suggested
  - convert LSM303D instead of removing it - tested on raspilot
  - minor improvements to LSM303D

**Comment by lucasdemarchi** [comment]
@guludo not sure if there's a race in the build system (linking stage arriving before the objects are built) but this error seems to come and go in semaphoreci:
https://semaphoreci.com/diydrones/ardupilot/branches/pull-request-3735/builds/8

I can't see how it would be related to these patches

**Comment by guludo** [comment]
@lucasdemarchi Thanks, I'll take a look today

**Comment by guludo** [comment]
Is that the correct link? I can't see the error there

**Comment by OXINARF** [comment]
@guludo It was the correct link because I saw the error too, but Semaphore probably relaunched the build. Here is one with the same error: https://semaphoreci.com/diydrones/ardupilot/branches/pull-request-3758/builds/5  I've saved the log in case it disappears again.

File: libraries/AP_Compass/AP_Compass_LSM303D.cpp [hunk 9]
```diff
@@ -236,6 +236,11 @@ bool AP_Compass_LSM303D::_read_sample()
         return false;
     }
 
+    /* check for overrun */
+    if ((rx.status & 0x70) != 0) {
```
Comment (guludo) [inline libraries/AP_Compass/AP_Compass_LSM303D.cpp]: No macros, huh. Then I looked at the whole file and there are several places like this. Ok, I think we can let that to another PR.

**Comment by guludo** [comment]
@lucasdemarchi thanks for the rework.

Although some of my comments weren't addressed as I expected, I think it looks good.

**Comment by lucasdemarchi** [comment]
Applied, thanks all for reviewing.
